lib/jxl/render_pipeline/simple_render_pipeline.cc

```diff
@@ -6,16 +6,146 @@
-  // TODO(veluca): actually run the pipeline.
+  if (!IsDone()) return;
+  for (size_t stage_id = 0; stage_id < stages_.size(); stage_id++) {
+    const auto& stage = stages_[stage_id];
+    // Prepare buffers for kInOut channels.
+    std::vector<ImageF> new_channels(channel_data_.size());
+    std::vector<ImageF*> output_channels(channel_data_.size());
+
+    std::vector<std::pair<size_t, size_t>> input_sizes(channel_data_.size());
+    for (size_t c = 0; c < channel_data_.size(); c++) {
+      input_sizes[c] =
+          std::make_pair(channel_data_[c].xsize() - kRenderPipelineXOffset * 2,
+                         channel_data_[c].ysize() - kRenderPipelineXOffset * 2);
+    }
+
+    for (size_t c = 0; c < channel_data_.size(); c++) {
+      if (stage->GetChannelMode(c) != RenderPipelineChannelMode::kInOut) {
+        continue;
+      }
+      new_channels[c] =
+          ImageF((input_sizes[c].first << stage->settings_.shift_x) +
+                     kRenderPipelineXOffset * 2,
+                 (input_sizes[c].second << stage->settings_.shift_y) +
+                     kRenderPipelineXOffset * 2);
+      output_channels[c] = &new_channels[c];
+    }
+
+    // Add mirrored pixes to all kInOut channels.
+    for (size_t c = 0; c < channel_data_.size(); c++) {
+      if (stage->GetChannelMode(c) != RenderPipelineChannelMode::kInOut) {
+        continue;
+      }
+      // Horizontal mirroring.
+      for (size_t y = 0; y < input_sizes[c].second; y++) {
+        float* row = channel_data_[c].Row(y + kRenderPipelineXOffset);
+        for (size_t ix = 0; ix < stage->settings_.border_x; ix++) {
+          row[kRenderPipelineXOffset - ix - 1] =
+              row[kRenderPipelineXOffset + ix];
+          row[kRenderPipelineXOffset + input_sizes[c].first + ix] =
+              row[kRenderPipelineXOffset + input_sizes[c].first - ix - 1];
+        }
+      }
+      // Vertical mirroring.
+      for (size_t iy = 0; iy < stage->settings_.border_y; iy++) {
+        memcpy(channel_data_[c].Row(kRenderPipelineXOffset - iy - 1) +
+                   kRenderPipelineXOffset - stage->settings_.border_x,
+               channel_data_[c].Row(kRenderPipelineXOffset + iy) +
+                   kRenderPipelineXOffset - stage->settings_.border_x,
+               sizeof(float) *
+                   (input_sizes[c].first + 2 * stage->settings_.border_x));
+      }
+    }
+
+    // All non-ignored channels should have the same size.
+    constexpr size_t kInf = size_t(-1);
+    size_t ysize = kInf;
+    size_t xsize = kInf;
+    for (size_t c = 0; c < channel_data_.size(); c++) {
+      if (stage->GetChannelMode(c) == RenderPipelineChannelMode::kIgnored) {
+        continue;
+      }
+      JXL_ASSERT(ysize == input_sizes[c].second || ysize == kInf);
+      ysize = input_sizes[c].second;
+      JXL_ASSERT(xsize == input_sizes[c].first || xsize == kInf);
+      xsize = input_sizes[c].first;
+    }
+
```

Comment:
Maybe have an assert here that xsize/ysize are not `kInf` anymore at this point? Better to catch cases where all channels are somehow ignored early, and not go do 2^64 iterations of a no-op loop or something.